[MRG+2] Classifier chain

adamklec opened the pull request:
This PR implements Classifier Chains for multi-label classification as described here:
http://www.cs.waikato.ac.nz/ml/publications/2009/chains.pdf
#3727 is a similar PR that seems to be stalled. This PR has one significant improvement over #3727. Since in general it is impossible to know the optimal ordering of classifiers for a particular problem a priori it is common to use an ensemble of randomly ordered classifier chains (as explained in the above paper). This implementation of classifier chains supports randomizing the chain order.

I included an example that demonstrates that on a multi-label dataset an ensemble of 10 randomly ordered classifier chains out-performs independently trained models.

Commit:
initial commit

Commit:
added shuffle parameter

Commit:
fixes for python 3 support

Commit:
fixed formatting issues

Commit:
more formatting fixes

amueller commented:
The appveyor test failure seems related.

amueller commented on examples/multi_label/classifier_chain.py:
```diff
@@ -0,0 +1,39 @@
+import numpy as np
```
can you create an example that has a file name starting with ``plot_`` and plots things? Plots make for more engaging examples, and only examples that start with ``plot_`` are automatically run. Also, if you want to add a new folder to the examples, you need to create an empty README.txt file in that folder.

adamklec replied:
I will work on adding an example with plots. Also please see my comment below regarding the README.txt file.

amueller replied:
I was about to suggest the confusion matrix, but I think we haven't implemented the multi-label confusion matrix yet.
It's actually not really important to add a plot for it to be a ``plot_`` example. That just ensures it runs. But it requires that the example runs quickly enough. If it doesn't, maybe try yeast instead?

adamklec replied:
I switched to yeast. I like it. It's much faster (the example runs in a few seconds). And it makes for a more convincing example.

amueller commented:
there shouldn't be an ``__init__.py`` file in the examples folder, that might be the error.

amueller commented on sklearn/multi_label/classifier_chain.py:
```diff
@@ -0,0 +1,102 @@
+        self.classifiers = []
+        self.chain_order = None
+
+    def fit(self, X, Y, chain_order=None, shuffle=True):
```
hm I'm not sure if we use capital Y for labels (though it would be logical).

adamklec replied:
I'd be happy to lower case it or change it to anything else. Just LMK what the correct convention is.

adamklec replied:
For what it's worth the convention of capitalizing the variable names of multi-label arrays already appears in the documentation here:
http://scikit-learn.org/stable/auto_examples/plot_multilabel.html#sphx-glr-auto-examples-plot-multilabel-py

amueller commented on sklearn/multi_label/classifier_chain.py:
```diff
@@ -0,0 +1,102 @@
+    """
+
+    def __init__(self,
+                 base_estimator=LogisticRegression(),
```
I'm not sure if having a mutable default is a good idea.

adamklec replied:
Agreed. I got rid of it.

amueller commented:
Thanks, this looks interesting. Can you please also add an entry to the user guide? And maybe ping @arjoly ?

Commit:
removed default base_estimator

Commit:
doc string formatting
removed logistic regression dependency

adamklec commented:
I removed the __init__.py file from the examples folder and added an README.txt file. The following exception is now getting thrown on CircleCI.

> Exception occurred:
>   File "/home/ubuntu/scikit-learn/doc/sphinxext/sphinx_gallery/gen_rst.py", line 182, in get_docstring_and_rest
>     .format(filename))
> ValueError: Could not find docstring in file "../examples/multi_label/classifier_chain.py". A docstring is required by sphinx-gallery
> The full traceback has been saved in /tmp/sphinx-err-BTEGNb.log, if you want to report the issue to the developers.
> Please also report this if it was a user error, so that a better error message can be provided next time.
> A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!

I tried copying the format of the README.txt files in other /examples directories but that had no effect. Any ideas?

GaelVaroquaux commented:
Put a docstring at the beginning of your example with a title and a text
that describe what the example does.

adamklec commented:
Ah. Thanks @GaelVaroquaux !

Commit:
added doctoring to classifier_chain example

Commit:
added multi_label to setup.py

Commit:
bug fix

Commit:
updated classifier chain example

Commit:
formatting fix

adamklec commented:
How do I go about adding an entry to the user guide?

jnothman commented:
Edit `doc/modules/multiclass.rst`

amueller commented on examples/multi_label/classifier_chain.py:
```diff
@@ -0,0 +1,86 @@
+============================
+
+Demonstrate that a single classifier chain
+out performs 10 independent classifiers
```
``out-performs``?

amueller replied:
Also format so that it's about 79 char each line (hopefully your editor can do that for you)

amueller commented on examples/multi_label/classifier_chain.py:
```diff
@@ -0,0 +1,86 @@
+from sklearn.datasets import fetch_rcv1
+
+# Get the Reuters Corpus Volume I dataset
+rcv1 = fetch_rcv1()
```
Uh, can you actually check how long this thing runs? If it's too long, it doesn't qualify as a ``plot_`` example because it would make the doc build too slow.

adamklec replied:
Yeah. I was worried about that. With my connection I believe it took a couple minutes to download. Of course that time is depends on connection speed. But it doesn't seem like something that should be run automatically.

Previously I was using **make_multilabel_classification** to generate a fake dataset which is obviously much faster. But I don't think this works because I believe the labels in the fake dataset are uncorrelated which means there is no advantage to using a classifier chain.

amueller replied:
I was actually not talking about the download (though that's also an issue) but about the computation. Have you tried yeast? That should have some correlations. And if ``make_multilabel_classification`` is not helpful for multi-label classification we should change/fix it.

amueller commented on sklearn/multi_label/classifier_chain.py:
```diff
@@ -0,0 +1,98 @@
+    ----------
+    classifiers_ : array
+        List of classifiers, which will be used to chain prediction.
+    chain_order : list of ints
```
Anything that's estimated during training should have a trailing underscore.

jnothman replied:
I'd be okay with dropping `chain_` and just making this `order_`

amueller commented on sklearn/multi_label/classifier_chain.py:
```diff
@@ -0,0 +1,98 @@
+
+    """
+
+    def __init__(self,
```
on one line?

Commit:
now using numpy RandomState
example now uses the yeast dataset instead of reuters
formatting fixes

Commit:
random_state fix

Commit:
removed unnecessary import

Commit:
formatting fix

adamklec commented:
CircleCI seems to be timing out. Not sure why.

Commit:
comment fix

adamklec commented:
Not sure what's going on here. The build is succeeding on CircleCI but then it times out.

amueller commented:
Don't worry about it. There's something weird going on with CircleCI right now.

Commit:
minor comment fix

Commit:
formatting fix

adamklec commented:
The previous build passed on CircleCI but failed on TravisCI because of a formatting issue. I fixed the formatting issue but it's back to timing out on CircleCI.

amueller commented:
that's a general issue we have right now. Feel free to investigate but that's not really on you.

adamklec commented:
Just want to check in and to see if the CircleCI issues have been resolved. Should I try pushing again?

Commit:
pushing again

Commit:
formatting fix

Commit:
added documentation

Commit:
Merge branch 'classifier_chain' of https://github.com/adamklec/scikit-learn into classifier_chain

HimaVarsha94 commented:
Has this feature been pushed on successfully?

jnothman commented:
I have no idea what "pushed on successfully" means. It has been successfully pushed onto my reviewing heap.

jnothman commented on sklearn/multi_label/classifier_chain.py:
```diff
@@ -0,0 +1,97 @@
+
+        self.base_estimator = base_estimator
+        self.random_state = random_state
+        self.classifiers_ = []
```
Only set this in `fit`

jnothman commented on sklearn/multi_label/classifier_chain.py:
```diff
@@ -0,0 +1,97 @@
+        self.base_estimator = base_estimator
+        self.random_state = random_state
+        self.classifiers_ = []
+        self.chain_order = None
```
Only set `chain_order_` (note trailing underscore) in `fit`.

jnothman commented on sklearn/multi_label/classifier_chain.py:
```diff
@@ -0,0 +1,97 @@
+        self.classifiers_ = []
+        self.chain_order = None
+
+    def fit(self, X, Y, chain_order=None, shuffle=True):
```
`chain_order` and `shuffle` should be estimator parameters, not `fit` parameters.

jnothman commented on sklearn/multi_label/classifier_chain.py:
```diff
@@ -0,0 +1,97 @@
+        """
+        Parameters
+        ----------
+        shuffle : bool
```
`X` and `Y` should be described here

jnothman commented on sklearn/multi_label/classifier_chain.py:
```diff
@@ -0,0 +1,97 @@
+        chain_order: list
+            A list of integers specifying the order of the classes in the
+            chain.
+        """
```
Needs a Returns section

jnothman commented on sklearn/multi_label/classifier_chain.py:
```diff
@@ -0,0 +1,97 @@
+            chain.
+        """
+        random_state = check_random_state(self.random_state)
+        self.classifiers_ = [clone(self.base_estimator) for _ in range(
```
I find the following wrapping much more readable:
```python
        self.classifiers_ = [clone(self.base_estimator)
                             for _ in range(Y.shape[1])]
```

jnothman commented on sklearn/multi_label/classifier_chain.py:
```diff
@@ -0,0 +1,97 @@
+            Y.shape[1])]
+
+        if chain_order is not None:
+            assert len(chain_order) == Y.shape[1], "chain_order length must " \
```
If this is validating user input it should be raising a `ValueError`. Assertion errors out of tests are only used to check things like invariants are maintained, or that some transformation had expected behaviour.

jnothman commented on sklearn/multi_label/classifier_chain.py:
```diff
@@ -0,0 +1,97 @@
+        for chain_idx, classifier in enumerate(self.classifiers_):
+            previous_labels = Y[:, self.chain_order[:chain_idx]]
+            X_aug = np.hstack((X, previous_labels))
+            y = np.squeeze(Y[:, self.chain_order[chain_idx]])
```
why is `squeeze` used?

jnothman commented on sklearn/multi_label/tests/test_classifier_chain.py:
```diff
@@ -0,0 +1,17 @@
+                                          n_classes=10)
+    classifier_chain = ClassifierChain(LogisticRegression())
+    classifier_chain.fit(X, Y)
+    Y_pred = classifier_chain.predict(X)
```
Should be checking the values of `classifiers_` and `chain_order_`. Should be testing the setting of `chain_order` and `shuffle`.

Should be checking that `[c.coef_.size for c in classifiers_] == range(X.shape[1], X.shape[1] + Y.shape[1])`

Should probably also be checking for a minimum training accuracy.

jnothman reviewed (changes requested):
1. Is this technique more generally applicable to multioutput problems? (regression or classification)
2. We should ensure this supports `Y` as a sparse matrix in the multilabel case
3. This belongs in `sklearn/multioutput.py`
4. This needs narrative docs, apparently in `doc/modules/multiclass.rst`

adamklec commented:
@jnothman Thanks for your feedback. Working on your requested changes now.

Commit:
requested changes

Commit:
added doc file

Commit:
minor fixes. added support for sparse labels/

Commit:
updated doc strings.
renamed chain_order to order.
added support for sparse labels.
added tests.

adamklec commented:
@jnothman I just pushed a commit that addresses most of the changes you requested. Tests are failing though. I am unable to access the test results on AppVeyor so I'm not sure what the issue is there. 

On CircleCI I see the following error which I do not understand.

```Exception occurred:
  File "/home/ubuntu/scikit-learn/doc/sphinxext/sphinx_gallery/docs_resolv.py", line 48, in _get_data
    with open(url, 'r') as fid:
IOError: [Errno 2] No such file or directory: '/home/ubuntu/scikit-learn/doc/_build/html/stable/modules/generated/sklearn.feature_selection.SelectKBest.rst.html'
The full traceback has been saved in /tmp/sphinx-err-zrxIfJ.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
Embedding documentation hyperlinks in examples..
	processing: feature_stacker.html
make: *** [html] Error 1

./build_tools/circle/build_doc.sh returned exit code 2

Action failed: ./build_tools/circle/build_doc.sh```

jnothman commented:
Please merge the latest master into your branch (or rebase on it), and that
problem will go away.

On 6 January 2017 at 08:02, adamklec <notifications@github.com> wrote:

> @jnothman <https://github.com/jnothman> I just pushed a commit that
> addresses most of the changes you requested. Tests are failing though. I am
> unable to access the test results on AppVeyor so I'm not sure what the
> issue is there.
>
> On CircleCI I see the following error which I do not understand.
>
>   File "/home/ubuntu/scikit-learn/doc/sphinxext/sphinx_gallery/docs_resolv.py", line 48, in _get_data
>     with open(url, 'r') as fid:
> IOError: [Errno 2] No such file or directory: '/home/ubuntu/scikit-learn/doc/_build/html/stable/modules/generated/sklearn.feature_selection.SelectKBest.rst.html'
> The full traceback has been saved in /tmp/sphinx-err-zrxIfJ.log, if you want to report the issue to the developers.
> Please also report this if it was a user error, so that a better error message can be provided next time.
> A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
> Embedding documentation hyperlinks in examples..
> 	processing: feature_stacker.html
> make: *** [html] Error 1
>
> ./build_tools/circle/build_doc.sh returned exit code 2
>
> Action failed: ./build_tools/circle/build_doc.sh```
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7602#issuecomment-270756632>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz656bDGhyJiC51W78znQGxaFBmPqYks5rPVpagaJpZM4KRXOp>
> .
>

Commit:
update doc string

Commit:
Merge branch 'master2' into classifier_chain

# Conflicts:
#	sklearn/__init__.py

Commit:
removed main

Commit:
renamed multi_label to multilabel.
removed qda from sklearn/__init__.py list.

Commit:
Merge branch 'master2' into classifier_chain

jnothman commented:
Please let us know when this is ready for another review.

Commit:
convert range to list for python 3

Commit:
fix for passing range as order

Commit:
remove old doc strings

Commit:
fix sparse labels snd formatting

Commit:
sparse label fix

Commit:
label slicing fix

Commit:
added performance test and descriptions to test_classifier_chain

adamklec commented:
@jnothman I believe this is ready for another review

jnothman commented on sklearn/multilabel/classifier_chain.py:
```diff
@@ -0,0 +1,134 @@
+
+
+class ClassifierChain(BaseEstimator):
+    """
```
PEP257: a one line description belongs here

jnothman commented on sklearn/multilabel/classifier_chain.py:
```diff
@@ -0,0 +1,134 @@
+
+class ClassifierChain(BaseEstimator):
+    """
+    A multi-label model that arranges binary classifiers into a chain. Each
```
Really, a multi-output model that arranges single-output (binary or multiclass) classifiers into a chain?

adamklec replied:
I'm happy to change the wording of this docstring but I'm unclear on the definition of "multi-output".  My implementation follows the original paper which was specifically a multi-label algorithm.

jnothman commented on sklearn/multilabel/classifier_chain.py:
```diff
@@ -0,0 +1,134 @@
+            State or seed for random number generator.
+
+        order : list
+            A list of integers specifying the order of the classes in the
```
order of the outputs/targets?

jnothman commented on sklearn/multilabel/classifier_chain.py:
```diff
@@ -0,0 +1,134 @@
+
+    def __init__(self, base_estimator, random_state=None,
+                 order=None, shuffle=True):
+        """
```
This detail belongs in the class docstring, not `__init__`

jnothman commented on sklearn/multilabel/classifier_chain.py:
```diff
@@ -0,0 +1,134 @@
+        """
+        self.base_estimator = base_estimator
+        self.random_state = random_state
+        if order is not None:
```
Please perform this validation in `fit`, and not here.

jnothman commented on sklearn/multilabel/classifier_chain.py:
```diff
@@ -0,0 +1,134 @@
+        self.shuffle = shuffle
+
+    def fit(self, X, Y):
+        """Fit the ClassifierChain model according to the given training (X)
```
PEP257 wants a one-line description

jnothman commented on sklearn/multilabel/classifier_chain.py:
```diff
@@ -0,0 +1,134 @@
+            classifier.fit(X_aug, y)
+
+    def predict(self, X):
+        """For each row in X make a prediction for each of the models in
```
PEP257: one-liner

jnothman commented on sklearn/multilabel/classifier_chain.py:
```diff
@@ -0,0 +1,134 @@
+
+        Attributes
+        ----------
+        classifiers_ : list
```
Conventionally we only put attributes in the class docstring.

Please rename it to `estimators_`

jnothman commented on sklearn/multilabel/tests/test_classifier_chain.py:
```diff
@@ -0,0 +1,59 @@
+                                       order=range(10),
+                                       shuffle=True)
+    classifier_chain.fit(X, Y)
+    assert (classifier_chain.order != list(range(10)))
```
Please use the helpers from `sklearn.utils.testing`, such as `assert_not_equal`. They give more helpful error messages

jnothman commented on doc/modules/classifier_chain.rst:
"multi-label or multi-output"

jnothman commented on sklearn/multilabel/classifier_chain.py:
```diff
@@ -0,0 +1,134 @@
+        for chain_idx, classifier in enumerate(self.classifiers_):
+            previous_predictions = Y_pred_chain[:, :chain_idx]
+            X_aug = np.hstack((X, previous_predictions))
+            Y_pred_chain[:, chain_idx] = classifier.predict(X_aug)
```
For classification, we might want to be able to incorporate not just the categorical predictions of the previous model, but its `predict_proba` or `decision_function` downstream in the chain. We can leave this feature until version 2.0 (just as we might leave multioutput regression to version 2.0, but I'd like to see that)

adamklec replied:
I'm not familiar with `decision_function`. For a while I had an implementation for `predict_proba` but in my experiments it did not help so I got rid of it and strictly follow the algorithm described in the paper. It wouldn't be hard to put it back in but as you suggested, I think many of these variations can be left for version 2.0.

jnothman commented on doc/modules/classifier_chain.rst:
among targets

jnothman commented on doc/modules/classifier_chain.rst:
```diff
@@ -0,0 +1,32 @@
+
+.. currentmodule:: sklearn.classifier_chain
+
+Classifier chains are a way of combining a number of binary classifiers
```
binary -> binary or multiclass

jnothman replied:
Please reference the class with
```
:class:`ClassifierChain`
```
at some point.

jnothman commented on sklearn/multilabel/classifier_chain.py:
```diff
@@ -0,0 +1,134 @@
+        else:
+            self.order = list(range(Y.shape[1]))
+
+        if self.shuffle:
```
This only makes sense if `order` is None; maybe instead `order='random'` should be used.

jnothman commented on sklearn/multilabel/classifier_chain.py:
```diff
@@ -0,0 +1,134 @@
+        for chain_idx, classifier in enumerate(self.classifiers_):
+            previous_labels = Y[:, self.order[:chain_idx]]
+            y = Y[:, self.order[chain_idx]]
+            X_aug = np.hstack((X, previous_labels))
```
As with stacking (see #7427), it makes sense during fitting if we do not use the gold-standard labels, but predictions from jacknifed model, i.e. the output of `cross_val_predict`. Could you add an optional `cv` parameter to support this approach, as in #7427?

adamklec replied:
Interesting. In the original paper they used the true labels during fitting but I can see the argument for using predictions. I will take a look at #7427.

jnothman commented on sklearn/multilabel/classifier_chain.py:
```diff
@@ -0,0 +1,134 @@
+            for column 3, etc. If chain_order is not None it must have a length
+            equal to the number of columns in Y.
+            If chain_order is None an ordered list of integers will be used
+                chain_order = [0, 1, 2, ..., Y.shape[1]]
```
really the upper bound is `Y.shape[1] - 1`

jnothman reviewed (changes requested):
Thanks. I'd still like to see a number of extensions here for this to be most useful.

Please also test the more general multi-output case, even by just changing some multilabel samples' labels at random to 2.

Please move the code to sklearn/multioutput.py

Enhancement wish-list in priority order:
* [ ] support for multitarget classification (I think you've already got this, apart from `decision_function` and `predict_proba`)
* [x] provide `decision_function` and `predict_proba` if available on `base_estimator`; you may use `if_delegate_has_method`.
* [x] support for jacknifed predictions in fitting (with `cv`)
* [ ] support for multitarget regression (with a separate `RegressionChain`)
* [ ] support for chained features being outputs of `predict_proba` or `decision_function` not just `predict`

(And if you feel like this is too daunting and want someone else to help, just shout out.)

jnothman commented on sklearn/multilabel/classifier_chain.py:
```diff
@@ -0,0 +1,134 @@
+        random_state : int or RandomState, optional, default None
+            State or seed for random number generator.
+
+        order : list
```
make this an `array-like, shape (n_outputs,)`

and add tests to ensure that list/array are both supported.

jnothman commented on sklearn/multilabel/classifier_chain.py:
```diff
@@ -0,0 +1,134 @@
+                             for _ in range(Y.shape[1])]
+
+        if self.order is not None:
+            if not len(self.order) == Y.shape[1]:
```
The real criterion is that `sorted(self.order) == list(range(Y.shape[1]))`

jnothman commented on doc/modules/classifier_chain.rst:
```diff
@@ -0,0 +1,32 @@
+Classifier Chain
+================
+
+.. currentmodule:: sklearn.classifier_chain
```
This module does not exist. The appropriate place for documentation appears to be `doc/modules/multiclass.rst`

jnothman commented on doc/modules/classifier_chain.rst:
Please reference the example

jnothman commented on sklearn/multilabel/classifier_chain.py:
```diff
@@ -0,0 +1,134 @@
+    chain_order is not known a priori it is common to use the mean prediction
+    of an ensemble randomly ordered classifier chains.
+
+    References
```
Please add a doctest example.

adamklec commented on doc/modules/classifier_chain.rst:
Ok. Although again, I'm not clear on the definition of multi-output.

jnothman commented:
multi-output (or multi-target) is a generalisation of multi-label: learn multiple possibly-correlated multiclass classifiers from the same data. Most scikit-learn estimators that support one support both currently.

jnothman commented:
(But the multilabel case is a very common specialisation, which is why the original is likely to have focused onit.)

Commit:
requested changes

Commit:
minor fixes

Commit:
minor fixes

Commit:
minor fix

Commit:
merged multilabel into multiclass

Commit:
move ClassifierChain to multioutput

Commit:
renaming

Commit:
more renaming

Commit:
minor fixes

Commit:
formatting

adamklec commented:
@jnothman I implemented most of your requested changes. Although I didn't attempt to add any items on your "enhancement wish list". 

I would also like add another item to the wish list. It is common to average the predictions of many randomly ordered classifier chains. It would be nice if we could use the existing ensemble methods. BaggingClassifier seems like it would be a natural choice. However it seems that BaggingClassifier is not currently compatible with multioutput models (`y : array of shape = [n_samples]` not `Y : {array-like, sparse matrix}, shape (n_samples, n_classes)`) Perhaps this is already on a todo list somewhere. If not making ensemble methods compatible with multioutput models seems like a good idea to me.

adamklec commented:
If I understand correctly, multilabel is a special case of multi-output. So rather than creating a new multilabel module I moved everything inside the existing multioutput module.

adamklec commented:
For some reason the CircleCI is failing when it tries to build the documentation. I am uncertain why. I will attempt to debug this later but I would appreciate it if someone could point me in the right direction regarding this error.

jnothman commented:
CircleCI fail is not your problem. It's due to a changed dependency, and we're onto it.

I consider those first couple of wish-list items fairly essential, if that's okay :)

Commit:
added predict_proba and decision_function

Commit:
minor changes. performance test is failing.

adamklec commented:
Ok. So I've added `predict_proba` and `decision_function` to `ClassifierChain` by rather naively changing the changing the call to `predict` to `predict_proba` and `decision_function`. `predict_proba` seems to work fine (using logistic regression) however `decision_function` (with LinearSVC as the base estimator) has significantly worse performance (the `jaccard_similarity_score` is about ~0.25 instead of ~0.5 for when using the regular `predict` method on LinearSVC or LogisticRegression). I'm not sure if this is due to a problem with my implementation, some aspect of `decision_function` I am not understanding correctly, or if for some reason the lower performance is to be expected.

Thoughts?

jnothman commented:
So you're chaining on `predict_proba` and `decision_function` at prediction time. Calling `decision_function` or `predict_proba` should still chain on the output of `predict`.

Chaining on `decision_function` will break it because the range of predictions is -infinity to +infinity, unlike `predict` which I think is {0, 1} and `predict_proba`, which ranges from 0 to 1.

Please make sure you test these for multiclass-multioutput too. It will require some changes. See, for instance, the description of `RandomForestClassifier.predict_proba`.

Something else I've realised you're missing is a `classes_` attribute. It is mandatory for any classifier.

Commit:
use binary predictions as features in decision_function and predict_proba

adamklec commented:
Using the binary predictions as features in `decision_function` resolved the performance issue. I made a similar change in `predict_proba`. Thanks for the suggestion.

What is the correct way to set the `classes_` attribute? Could it be something trivial like `self.classes = np.arange(len(self.estimators_))` or `self.classes = np.arange(len(Y.shape[1]))` ?

jnothman commented:
```
classes_ = [est.classes_ for est in self.estimators_]
```
I think... Best to check what `RandomForestClassifier` does with multilabel etc. in case I'm muddling things.

Commit:
formatting fix

Commit:
added classes_ attribute

adamklec commented:
@jnothman I added the classes_ attribute. Please review again at your convenience.

jnothman commented on sklearn/multioutput.py:
```diff
@@ -359,3 +361,187 @@ def score(self, X, y):
+        for chain_idx, estimator in enumerate(self.estimators_):
+            previous_labels = Y[:, self.order[:chain_idx]]
+            y = Y[:, self.order[chain_idx]]
+            X_aug = np.hstack((X, previous_labels))
```
I think you should preallocate `X_aug = np.zeros((X.shape[0], X.shape[1] + Y.shape[1]), dtype=...)` (in the dense, binary case), and then fit with `X_aug[:, X.shape[1] + chain_idx:]`

adamklec replied:
I like this.

jnothman commented on sklearn/multioutput.py:
```diff
@@ -359,3 +361,187 @@ def score(self, X, y):
+        if sp.issparse(Y):
+            Y = Y.toarray()
+
+        if sp.issparse(X):
```
This isn't acceptable. We assume that if sparse X is used, the memory consumption of dense X is infeasbile.

The same is true of `Y`, but sparse `Y` is not as common. Deal with this first.

adamklec replied:
I fixed the handling of sparse data and removed support for sparse labels (for now?) but I'm running into a problem where scipy raises an exception when you try to hstack an empty sparse array. This only seems to occur on some versions of python. Looking into possible solutions. Suggestions welcome.

jnothman reviewed (commented):
The `predict_proba`/`decision_function` logic looks like it should work for the multilabel case (I've not checked tests yet), but you'll need to change things to handle the multioutput multiclass case... perhaps later.

Commit:
optimization for sparse data. removed support for sparse labels.

Commit:
formatting fix

Commit:
sparse data fix

jnothman reviewed (commented):
I've been meaning to give this a more complete review, but here are some nitpicks that have been sitting around for days.

jnothman reviewed (commented):
I'd really like to see the option to use jackknifed predictions (`cross_val_predict`), rather than gold standard Y, in `fit`. This is just a matter of having a `y_pred_chain` at fit time, and setting each column in turn to `y if self.cv is None else cross_val_predict(self.base_estimator, X_aug, y, cv=self.cv)`. I'll give this a full review once that's implemented, and would be willing to give it my approval without support for the multioutput-multiclass and multioutput-regressor generalisations (though these would be great!).

adamklec commented:
@jnothman I've been out/busy for the last week or so but I hope to take another pass at this soon.

Commit:
requested changes

Commit:
merge master into branch

Commit:
sparse fix

Commit:
added support for cross_val_predict to ClassifierChain

Commit:
formatting fix

jnothman commented on sklearn/multioutput.py:
```diff
@@ -359,3 +362,229 @@ def score(self, X, y):
+                    if chain_idx == 0:
+                        X_aug = X
+                    else:
+                        X_aug = sp.hstack((X, Y[:, self.order[:chain_idx]]))
```
This should be `Y_pred_chain`

jnothman commented on sklearn/multioutput.py:
```diff
@@ -359,3 +362,229 @@ def score(self, X, y):
+        self.classes_ = []
+
+        if sp.issparse(X):
+            if self.cv is None:
```
Is it possible to factor this so the if/else is minimal, i.e. something like:
```
if self.cv is None:
    Y_pred_chain = y
else:
    Y_pred_chain = np.zeros(...)
if chain_idx == 0:
...
estimator.fit(X_aug, y)
self.classes_.append(estimator.classes_)
if self.cv is not None:
    Y_pred_chain[:, chain_idx] = cross_val_predict(...)

jnothman commented on sklearn/multioutput.py:
```diff
@@ -359,3 +362,229 @@ def score(self, X, y):
+                    self.classes_.append(estimator.classes_)
+
+            else:
+                Y_pred_chain = np.zeros((X.shape[0], len(self.estimators_)))
```
You could again avoid doing the hstack in the loop by hstacking zeros onto X as above with X and Y

jnothman commented on sklearn/multioutput.py:
```diff
@@ -359,3 +362,229 @@ def score(self, X, y):
+        Y_pred : array-like, shape (n_samples, n_classes)
+        """
+
+        Y_pred_chain = np.zeros((X.shape[0], len(self.estimators_)))
```
You can again hstack X to this and fill it in as you go to avoid repeated copying.

jnothman commented on sklearn/tests/test_multioutput.py:
```diff
@@ -336,3 +341,123 @@ def test_multi_output_exceptions():
+    classifier_chain = ClassifierChain(LogisticRegression())
+    classifier_chain.fit(X, Y)
+    Y_pred = classifier_chain.predict(X)
+    Y_pred_binary = (Y_pred >= .5)
```
Isn't it already binary?

jnothman commented on sklearn/tests/test_multioutput.py:
```diff
@@ -336,3 +341,123 @@ def test_multi_output_exceptions():
+
+def test_classifier_chain_fit_and_predict():
+    """Fit classifier chain and verify predict performance"""
+    X, Y = datasets.make_multilabel_classification(n_samples=10000,
```
Why do we want so many samples? we need tests to run quickly.

adamklec replied:
The generated dataset has 10 classes. `cross_val_predict` will throw errors if one of the cv splits doesn't have a positive example for one of the classes. This happens rather often with N=1000 but I've never seen it happen with N=10000

jnothman replied:
That may be a bug in `cross_val_predict`. However:
* it may be appropriate to construct or use a multilabel dataset in another way, particularly if our validation that it is working goes beyond "jaccard > 0.5?" which ideally it should
* I don't see why you need `n_classes=10`. Surely this proves nothing more than `n_classes=3`
* `n_labels` controls the density of labels. Increasing it will mean you will very likely have positive examples for smaller `n_samples`.

jnothman replied:
Note that `make_multilabel_classification` is not a very good dataset for classifier chains: there is no covariance between labels imposed by the model.

adamklec replied:
That is a very good point about `make_multilabel_classification` not having covariances. I was wondering about that. You're right that it doesn't make much sense to use such a dataset for performance tests. I could switch to using the yeast dataset for all of these tests.

jnothman commented on sklearn/tests/test_multioutput.py:
```diff
@@ -336,3 +341,123 @@ def test_multi_output_exceptions():
+    classifier_chain = ClassifierChain(LogisticRegression(), cv=3)
+    classifier_chain.fit(X, Y)
+    Y_pred = classifier_chain.predict(X)
+    Y_pred_binary = (Y_pred >= .5)
```
I think we should at least check that the models with `cv=None` and `cv=3` are different

jnothman commented on sklearn/tests/test_multioutput.py:
```diff
@@ -336,3 +341,123 @@ def test_multi_output_exceptions():
+                                                   n_classes=10)
+    classifier_chain = ClassifierChain(LinearSVC())
+    classifier_chain.fit(X, Y)
+    Y_pred = classifier_chain.predict(X)
```
I think it's okay to test `predict`, `predict_proba`, `decision_function` and `coef_` sizes all in one test function.

Commit:
requested changes

jnothman commented:
Could you just use make_classification and convert, say 4 or 6 classes to
some binary encoding?

On 8 Mar 2017 11:34 am, "Adam Kleczewski" <notifications@github.com> wrote:

> *@adamklec* commented on this pull request.
> ------------------------------
>
> In sklearn/tests/test_multioutput.py
> <https://github.com/scikit-learn/scikit-learn/pull/7602#discussion_r104817689>
> :
>
> > @@ -336,3 +341,123 @@ def test_multi_output_exceptions():
>      y_new = np.column_stack((y1, y2))
>      moc.fit(X, y)
>      assert_raises(ValueError, moc.score, X, y_new)
> +
> +
> +def test_classifier_chain_fit_and_predict():
> +    """Fit classifier chain and verify predict performance"""
> +    X, Y = datasets.make_multilabel_classification(n_samples=10000,
>
> That is a very good point about make_multilabel_classification not having
> covariances. I was wondering about that. You're right that it doesn't make
> much sense to use such a dataset for performance tests. I could switch to
> using the yeast dataset for all of these tests.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7602#discussion_r104817689>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz64gIqflpCr2uhO_4ZLhGEjDZUQw3ks5rjfetgaJpZM4KRXOp>
> .
>

adamklec commented:
Not sure I understand what you're suggesting...

jnothman commented:
use data consisting of 4 classes, for instance, and map:
0 -> 0, 0
1 -> 0, 1
2 -> 1, 0
3 -> 1, 1

thus it is now a multilabel problem. The labels are still conditionally
independent given X, but there is an issue of integrity requiring the
classifiers to agree.

On 9 March 2017 at 03:25, Adam Kleczewski <notifications@github.com> wrote:

> Not sure I understand what you're suggesting...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7602#issuecomment-285089467>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz688_ivi-fKZO73wbKxZwbXpFvJRaks5rjtZjgaJpZM4KRXOp>
> .
>

Commit:
New dataset for testing ClassifierChain.
Fixes to classifier chain yeast example.
Bug fix in ClassifierChain cross_val_predict

adamklec commented:
Got it. That seems to work although the benefits of using a classifier chain on datasets generated in the way you described is much smaller than on the yeast dataset (compared to independent models).

adamklec commented:
I changed the tests to use datasets generated in the way you described above.

jnothman commented:
Yeast, then, may be appropriate for an example. And perhaps for one
performance-based test

On 10 Mar 2017 5:09 am, "Adam Kleczewski" <notifications@github.com> wrote:

> I changed the tests to use datasets generated in the way you described
> above.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7602#issuecomment-285432198>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz676OuFzVt1vqr3QwUvbLNzUL_GQ7ks5rkEBggaJpZM4KRXOp>
> .
>

Commit:
added a performance test on the yeast dataset

jnothman commented on sklearn/tests/test_multioutput.py:
```diff
@@ -336,3 +345,158 @@ def test_multi_output_exceptions():
+    for chain in chains:
+        chain.fit(X_train, Y_train)
+
+    Y_pred_ensemble = np.array([chain.predict(X_test) for chain in
```
Why an ensemble? For a test you can just hard-code an order that works well.

adamklec replied:
Fair point. It doesn't need to be an ensemble. I changed it to a single chain with a fixed order.

I haven't looked into how one might go about finding the optimal chain ordering. When I've used these chains in my work I always just used an ensemble of randomly ordered chains. For about 5 minutes I thought about using a genetic algorithm to find the optimal chain ordering but I never tried it.

jnothman reviewed (commented):
By the way, is there literature on choosing a good order without cross validation?

Commit:
changed performance test on yeast dataset to use a single chain with a fixed order and a non-random train/test split

adamklec commented:
I haven't looked into how one might go about finding the optimal chain ordering. When I've used these chains in my work I always just used an ensemble of randomly ordered chains. For about 5 minutes I thought about using a genetic algorithm to find the optimal chain ordering but I never tried it.

Commit:
removed train test split import from classifier chain performance test

jnothman commented on sklearn/multioutput.py:
```diff
@@ -359,3 +362,218 @@ def score(self, X, y):
+    random_state : int or RandomState, optional, default None
+        State or seed for random number generator.
+
+    order : list of integers, None, or 'random'
```
I think type should be array-like, shape=[n_outputs] or 'random', optional

jnothman commented on sklearn/multioutput.py:
```diff
@@ -359,3 +362,218 @@ def score(self, X, y):
+        column 1 in the Y matrix, the second model will make predictions
+        for column 3, etc.
+
+        If order is None the order will be determined by the
```
I would prefer "by default"

jnothman commented on sklearn/multioutput.py:
```diff
@@ -359,3 +362,218 @@ def score(self, X, y):
+            else:
+                X_aug = np.hstack((X, previous_predictions))
+            Y_pred_chain[:, chain_idx] = estimator.predict(X_aug)
+        chain_key = [np.where(self.order == i)[0][0]
```
I think this can be done with no.argsort or just
```python
inv_order = no.empty_like(order)
Inv_order [order] = no.arange(len(order))
```

jnothman commented on sklearn/multioutput.py:
```diff
@@ -359,3 +362,218 @@ def score(self, X, y):
+            Returns self.
+        """
+
+        random_state = check_random_state(self.random_state)
```
You should be using check_X_y somewhere

adamklec replied:
Ok. This reminds me of something else. What convention are we following upper/lower case in variable names? I typically use uppercase for matrices and lowercase for vectors. But in `multioutput.py` label matrices seem to be called `y` even when they are matrices.

jnothman reviewed (commented):
Sorry for the drop by drop review...

adamklec commented on sklearn/multioutput.py:
```diff
@@ -359,3 +362,218 @@ def score(self, X, y):
+        If order is 'random' a random ordering will be used.
+
+    cv : int, cross-validation generator or an iterable, optional
+        Determines whether to use cross validated predictions or true
```
Ok. Although I would point out that the original paper used exclusively binary true labels as features. Furthermore, in my experience using the cross validation predictions as features at training time consistently leads to worse performance. I don't have an explanation of this but it has been my experience. Nevertheless if you confirm that you still want this change I am willing to make it.

jnothman replied:
Well we can leave open the option of having a way to specify 'predictions of one model'. Fine as it is for now.

(And for the record, I found that jackknifed predictions could help, with an appropriately difficult dataset)

jmschrei replied:
It's interesting that you see true labels leading to better performance. I was initially with @jnothman that it doesn't really make sense to train on features that won't really be available at test time. However, since the paper suggests it, and if you've observed better performance using them, it's probably best to leave it how it is, with both options.

Commit:
requested changes

adamklec commented:
Reposting this question from above.

What convention are we following for upper/lower casing in variable names? I typically use uppercase for matrices and lowercase for vectors. But in `multioutput.py` label matrices seem to be called `y`.

Commit:
update doc string

Commit:
update random_state doc string

Commit:
formatting fix

jnothman commented:
Btw, it looks like Bayesian Classifier Chains (Zaragoza et al, IJCAI 2011) provides a method of determining the order. I say this only from reading http://meka.sourceforge.net/api-1.7/meka/classifiers/multilabel/BCC.html. In general it might be a good idea to familiarise yourself with meka's toolkit to understand what variants exist and how the implementation might be extended.

adamklec commented:
No doubt there are many variations out there and we could extend this work in any number of directions. But I believe we've met the requirements for v1.0. Is there anything remaining to be done before we can merge?

jnothman commented:
I've not formally finished reviewing. Is also like to try it out. I've been
very pressed for time. So have our other core devs. You will require two of
us to approve the contribution before it is merged. Sorry and thanks for
your patience.

On 15 Mar 2017 1:36 am, "Adam Kleczewski" <notifications@github.com> wrote:

> No doubt there are many variations out there and we could extend this work
> in any number of directions. But I believe we've met the requirements for
> v1.0. Is there anything remaining to be done before we can merge?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7602#issuecomment-286440571>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz615Lj3_S0HjF29gfFL80qBo4Si-pks5rlqXmgaJpZM4KRXOp>
> .
>

jnothman commented on sklearn/tests/test_multioutput.py:
```diff
@@ -336,3 +344,160 @@ def test_multi_output_exceptions():
+    classifier_chain.fit(X, Y)
+    Y_pred = classifier_chain.predict(X)
+    assert_equal(Y_pred.shape, Y.shape)
+    assert_greater(jaccard_similarity_score(Y, Y_pred), 0.4)
```
please perform the `predict_proba` test here too.

adamklec replied:
done

jnothman commented on sklearn/tests/test_multioutput.py:
```diff
@@ -336,3 +344,160 @@ def test_multi_output_exceptions():
+    classifier_chain.fit(X, Y)
+    Y_pred = classifier_chain.predict(X)
+    assert_equal(Y_pred.shape, Y.shape)
+    assert_greater(jaccard_similarity_score(Y, Y_pred), 0.4)
```
please perform the `decision_function` test here too.

jnothman commented on sklearn/multioutput.py:
```diff
@@ -359,3 +362,228 @@ def score(self, X, y):
+        random_state = check_random_state(self.random_state)
+
+        if self.order is None:
+            self.order = np.array(range(Y.shape[1]))
```
we do not overwrite parameters, by convention. Instead, store an attribute `order_`.

jnothman commented on sklearn/tests/test_multioutput.py:
```diff
@@ -336,3 +344,160 @@ def test_multi_output_exceptions():
+
+def test_classifier_chain_vs_independent_models():
+    """Verify that an ensemble of classifier chains (each of length
+    N) achieves a higher Jaccard similarity score than N independent models"""
```
achieves -> can achieve

jnothman commented on sklearn/tests/test_multioutput.py:
```diff
@@ -336,3 +344,160 @@ def test_multi_output_exceptions():
+
+
+def test_classifier_chain_vs_independent_models():
+    """Verify that an ensemble of classifier chains (each of length
```
due to issues with nose, we don't usually use docstrings in test functions. Comments are preferred

jnothman reviewed (commented):
Thanks for your work, @adamklec. A complete review of your work so far.
* We should ideally support sparse `Y`, or make sure an error is thrown for now if sparse `Y` is passed.
* I [tried this out on the bibtex dataset](https://gist.github.com/jnothman/7af35286c61daac570a307ea0568eb5a) from http://mulan.sourceforge.net/datasets-mlc.html. Independent models gets 0.8240. `ClassifierChain(cv=None, order='random')` consistently gets ~0.822. `ClassifierChain(cv=None, order='random')` gets 0.8242 to 0.8246. So not great improvement, but a clear case where we're losing performance by fitting on true labels.

adamklec commented:
Thanks @jnothman. I've been busy the last few weeks but I will make these changes soon.

Commit:
requested changes

adamklec commented on sklearn/multioutput.py:
```diff
@@ -359,3 +362,228 @@ def score(self, X, y):
+                else:
+                    X_aug[:, col_idx] = cv_result
+
+            self.classes_.append(estimator.classes_)
```
done

Commit:
Merge branch 'master2' into classifier_chain

Commit:
formatting

adamklec commented on sklearn/multioutput.py:
```diff
@@ -359,3 +362,228 @@ def score(self, X, y):
+    ----------
+    Jesse Read, Bernhard Pfahringer, Geoff Holmes, Eibe Frank, "Classifier
+    Chains for Multi-label Classification", 2009.
+
```
done

Commit:
bug fixes

Commit:
added a plot to the yeast example

Commit:
formatting

Commit:
formatting

Commit:
formatting

Commit:
formatting

Commit:
formatting

Commit:
typo fixes

Commit:
bug fix

adamklec commented:
@jnothman I made the changes you requested. I also made significant changes to the example. Please have a look at your convenience.

The reason this is not green is because there seems to be a problem with the test that verifies that a classifier chain out performs a one vs rest classifier (`test_classifier_chain_vs_independent_models`). This is the only test that uses the yeast dataset instead of fake data. The fetch_mldata method seems to be failing inconsistently. The problem could be with  the network connectivity of the servers running the tests. Not sure. Any ideas?

jnothman commented:
mldata has been unreliable of late. thanks for your work. I'll try to get
to it soon but my plate is very full

On 3 May 2017 3:57 am, "Adam Kleczewski" <notifications@github.com> wrote:

> @jnothman <https://github.com/jnothman> I made the changes you requested.
> I also made significant changes to the example. Please have a look at your
> convenience.
>
> The reason this is not green is because there seems to be a problem with
> the test that verifies that a classifier chain out performs a one vs rest
> classifier (test_classifier_chain_vs_independent_models). This is the
> only test that uses the yeast dataset instead of fake data. The
> fetch_mldata method seems to be failing inconsistently. The problem could
> be with the network connectivity of the servers running the tests. Not
> sure. Any ideas?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7602#issuecomment-298712048>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6-rK-aZOWSRrNFT5PA_5Dtut4LPEks5r126IgaJpZM4KRXOp>
> .
>

jnothman commented on sklearn/multioutput.py:
```diff
@@ -359,3 +362,224 @@ def score(self, X, y):
+            Y_pred_chain = np.zeros((X.shape[0], len(self.estimators_)))
+
+        if sp.issparse(X):
+            X_aug = sp.hstack((X, Y_pred_chain), format='lil')
```
I get why lil is good (sort of) from the perspective of inserting successive predictions. But we usually benefit from CSR in predictors. In the cv is None case, therefore, CSR is much preferred. In the other case, we could do CSR too, but efficient insertion into an empty column takes a few lines of code; CSC would be a little more appropriate. You needn't worry about that case for now.

jnothman commented on sklearn/multioutput.py:
```diff
@@ -359,3 +362,224 @@ def score(self, X, y):
+        if self.cv is None:
+            Y_pred_chain = Y[:, self.order_]
+        else:
+            Y_pred_chain = np.zeros((X.shape[0], len(self.estimators_)))
```
if X is sparse, this array is being allocated for nothing. Better off with `lil_matrix((X.shape[0], Y.shape[1]))`

jnothman commented on sklearn/multioutput.py:
```diff
@@ -359,3 +362,224 @@ def score(self, X, y):
+        The original order of labels (as specified by Y) is recovered in
+        the predict method by indexing into the predictions with the
+        list of indices in the order attribute.
+
```
order_ is not documented.

jnothman commented on sklearn/multioutput.py:
```diff
@@ -359,3 +362,224 @@ def score(self, X, y):
+    estimators_ : list
+        A list of clones of base_estimator. Once fit the estimators in
+        this list will be ordered as specified by the order attribute.
+        The original order of labels (as specified by Y) is recovered in
```
I think this is extraneous detail.

jnothman commented on sklearn/multioutput.py:
```diff
@@ -359,3 +362,224 @@ def score(self, X, y):
+
+    Attributes
+    ----------
+    classes_ : list of arrays of length len(estimators_) containing the
```
the first line here should only contain the name and the type spec. The next line should describe.

jnothman commented on sklearn/multioutput.py:
```diff
@@ -359,3 +362,224 @@ def score(self, X, y):
+
+        The order of the chain can be explicitly set by providing a list of
+        integers. For example, for a chain of length 5
+            order = [1, 3, 2, 4, 0]
```
formatting.

jnothman commented on examples/multioutput/plot_classifier_chain_yeast.py:
I'm not sure, but I wonder if we should have `random_state=i` so this is replicable

jnothman commented on sklearn/multioutput.py:
```diff
@@ -359,3 +362,224 @@ def score(self, X, y):
+
+    order : array-like, shape=[n_outputs] or 'random', optional
+        By default the order will be determined by the order of columns in
+        the label matrix Y.
```
You need :: at the end of this line and a blank line for the following to be formatted as code. Otherwise it is formatted as a continuation of this line.

jnothman commented on sklearn/multioutput.py:
```diff
@@ -359,3 +362,224 @@ def score(self, X, y):
+            X_aug = sp.hstack((X, Y_pred_chain), format='lil')
+        else:
+            X_aug = np.hstack((X, Y_pred_chain))
+
```
del Y_pred_chain could be done here to free up a little memory

jnothman commented on sklearn/multioutput.py:
```diff
@@ -359,3 +362,224 @@ def score(self, X, y):
+        -------
+        Y_pred : array-like, shape (n_samples, n_classes)
+        """
+        Y_pred_chain = np.zeros((X.shape[0], len(self.estimators_)))
```
Need to do check_array before X.shape. We need to test that this works with lists of lists. For accept_sparse, CSC may be appropriate.

jnothman commented on sklearn/multioutput.py:
```diff
@@ -359,3 +362,224 @@ def score(self, X, y):
+        -------
+        Y_prob : array-like, shape (n_samples, n_classes)
+        """
+        Y_prob_chain = np.zeros((X.shape[0], len(self.estimators_)))
```
need check_array

jnothman commented on sklearn/multioutput.py:
```diff
@@ -359,3 +362,224 @@ def score(self, X, y):
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape (n_samples, n_features)
+        Y : {array-like, sparse matrix}, shape (n_samples, n_classes)
```
I don't think sparse Y is currently tested

adamklec commented:
I am calling `check_X_y` first thing in the fit method. Is it still necessary to call `check_array` later?

Commit:
requested changes

Commit:
make example reproducible

Commit:
formatting

Commit:
added check_array to predict methods

adamklec commented:
@jnothman I believe I have addressed all of your requested changes.

jnothman commented on sklearn/multioutput.py:
```diff
@@ -359,3 +362,234 @@ def score(self, X, y):
+        if self.cv is None:
+            Y_pred_chain = Y[:, self.order_]
+            if sp.issparse(X):
+                X_aug = sp.hstack((X, Y_pred_chain), format='lil')
```
I'd rather CSR or CSC format here if possible

adamklec replied:
If I just naively make the change I get this warning
`SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.
  SparseEfficiencyWarning)`
Is there a better way to use CSR or CSC here?

jnothman replied:
I only expected that warning when inserting into a CSC/CSR, and I thought we don't insert in the cv is None case

jnothman replied:
I only expected that warning when inserting into a CSC/CSR, and I thought we don't insert in the cv is None case

adamklec replied:
Ah. I think I misunderstood what you were asking for. LIL is ok when `self.cv is not None`?

jnothman commented on sklearn/multioutput.py:
```diff
@@ -359,3 +362,234 @@ def score(self, X, y):
+    def predict_proba(self, X):
+        """Predict probability estimates.
+
+        Probability estimates will be used as features by subsequent models
```
Not true

jmschrei replied:
Not true in the current implementation-- but should it be? Taking in probabilities might be advantageous.

jnothman commented on sklearn/multioutput.py:
```diff
@@ -359,3 +362,234 @@ def score(self, X, y):
+    def decision_function(self, X):
+        """Distance of the samples X to the separating hyperplane.
+
+        Decision function values will be used as features by subsequent
```
Not true

jnothman commented on sklearn/multioutput.py:
```diff
@@ -359,3 +362,234 @@ def score(self, X, y):
+
+    @if_delegate_has_method('base_estimator')
+    def decision_function(self, X):
+        """Distance of the samples X to the separating hyperplane.
```
I don't know that this is accurate for all decision_function implementations.

More accurate would be to define this in terms of the same method of the classifiers in the chain

jnothman reviewed (commented):
I've not yet done a final review of tests

jnothman commented on sklearn/tests/test_multioutput.py:
```diff
@@ -336,3 +344,131 @@ def test_multi_output_exceptions():
+    classifier_chain = ClassifierChain(LogisticRegression(),
+                                       order='random')
+    classifier_chain.fit(X, Y)
+    assert_not_equal(list(classifier_chain.order), list(range(4)))
```
Without using a fixed random seed, this will fail every 1/24 runs

adamklec replied:
good point

jnothman commented on sklearn/tests/test_multioutput.py:
```diff
@@ -336,3 +344,131 @@ def test_multi_output_exceptions():
+    classifier_chain.fit(X, Y)
+    assert_not_equal(list(classifier_chain.order), list(range(4)))
+    assert_equal(len(classifier_chain.order_), 4)
+    assert_equal(len(set(classifier_chain.order_)), 4)
```
Could even check that fitting a ClassifierChain with this order produces an identical model.

jnothman reviewed (approved):
Otherwise, LGTM!

Please fix those things up, and we'll await another review. Also please add an entry to doc/whats_new.rst

Thanks

jmschrei commented on examples/multioutput/plot_classifier_chain_yeast.py:
It might be nice if you put the expected performance results as a comment

jmschrei commented on sklearn/multioutput.py:
```diff
@@ -359,3 +362,234 @@ def score(self, X, y):
+        If None, the random number generator is the RandomState instance used
+        by `np.random`.
+
+        Used to generate random chain orders.
```
This should go up two lines, as the last line in the random_state entry.

jmschrei commented on sklearn/multioutput.py:
```diff
@@ -359,3 +362,234 @@ def score(self, X, y):
+        possible inputs for cv are:
+            * integer, to specify the number of folds in a (Stratified)KFold,
+            * An object to be used as a cross-validation generator.
+            * An iterable yielding train, test splits.
```
Specify the default somewhere here.

jmschrei commented on sklearn/multioutput.py:
```diff
@@ -359,3 +362,234 @@ def score(self, X, y):
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape (n_samples, n_features)
```
Is this the proper way to specify parameters? It seems different to me.

adamklec replied:
There seems to be some inconsistency in the code between `shape (n_samples, n_features)` and `shape = (n_samples, n_features)`. Otherwise I believe it is correct. I will add a second line describing the purpose of `X` and `Y` (i.e "The input data." and "The target data.").

jmschrei replied:
What I meant is the {array-like, sparse matrix} in parentheses. I think it should be `X : array-like or sparse matrix, shape = (n_samples, n_featues)`. Nitpick, though.

jmschrei commented on sklearn/tests/test_multioutput.py:
```diff
@@ -336,3 +344,131 @@ def test_multi_output_exceptions():
+
+    Y_pred = classifier_chain.predict(X)
+    assert_equal(Y_pred.shape, Y.shape)
+    assert_greater(jaccard_similarity_score(Y, Y_pred), 0.4)
```
How do you know this will be greater than 0.4?

adamklec replied:
A priori I didn't. I just picked a value that it consistently exceeded. I did this before I added the performance test on the iris dataset. But now that we are testing performance elsewhere this test is arguably not necessary. Unless I hear otherwise I will remove it.

jmschrei commented on sklearn/tests/test_multioutput.py:
```diff
@@ -336,3 +344,131 @@ def test_multi_output_exceptions():
+    assert_greater(jaccard_similarity_score(Y, Y_pred), 0.4)
+
+    Y_prob = classifier_chain.predict_proba(X)
+    Y_binary = (Y_prob >= .5)
```
+1

jmschrei commented:
This is a great addition. Thanks for taking the time to implement it and deal with the long review process. Hopefully we're close to getting this merged!

jnothman commented:
i found that cv improved performance, but you obviously need a problem
where overfitting is a big issue.

On 2 Jun 2017 2:22 pm, "Jacob Schreiber" <notifications@github.com> wrote:

*@jmschrei* commented on this pull request.
------------------------------

In sklearn/multioutput.py
<https://github.com/scikit-learn/scikit-learn/pull/7602#discussion_r119779119>
:

> +    order : list of integers, None, or 'random'
+        The order of the chain can be explicitly set by providing a list of
+        integers. For example, for a chain of length 5
+            order = [1, 3, 2, 4, 0]
+        means that the first model in the chain will make predictions for
+        column 1 in the Y matrix, the second model will make predictions
+        for column 3, etc.
+
+        If order is None the order will be determined by the
+        order of columns in the label matrix Y.
+            order = [0, 1, 2, ..., Y.shape[1] - 1]
+
+        If order is 'random' a random ordering will be used.
+
+    cv : int, cross-validation generator or an iterable, optional
+        Determines whether to use cross validated predictions or true

It's interesting that you see true labels leading to better performance. I
was initially with @jnothman <https://github.com/jnothman> that it doesn't
really make sense to train on features that won't really be available at
test time. However, since the paper suggests it, and if you've observed
better performance using them, it's probably best to leave it how it is,
with both options.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/scikit-learn/scikit-learn/pull/7602#discussion_r119779119>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/AAEz63nqCm2aiUJFbVVw_0JFU801tPbEks5r_44XgaJpZM4KRXOp>
.

jnothman commented:
see my wishlist above :)

On 2 Jun 2017 2:27 pm, "Jacob Schreiber" <notifications@github.com> wrote:

> *@jmschrei* commented on this pull request.
> ------------------------------
>
> In sklearn/multioutput.py
> <https://github.com/scikit-learn/scikit-learn/pull/7602#discussion_r119779432>
> :
>
> > +                    X_aug = sp.hstack((X, previous_predictions))
> +            else:
> +                X_aug = np.hstack((X, previous_predictions))
> +            Y_pred_chain[:, chain_idx] = estimator.predict(X_aug)
> +
> +        inv_order = np.empty_like(self.order_)
> +        inv_order[self.order_] = np.arange(len(self.order_))
> +        Y_pred = Y_pred_chain[:, inv_order]
> +
> +        return Y_pred
> +
> +    @if_delegate_has_method('base_estimator')
> +    def predict_proba(self, X):
> +        """Predict probability estimates.
> +
> +        Probability estimates will be used as features by subsequent models
>
> Not true in the current implementation-- but should it be? Taking in
> probabilities might be advantageous.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7602#discussion_r119779432>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz60XASCIkqkc3Yt5bngHE_gwwB_t0ks5r_484gaJpZM4KRXOp>
> .
>

jnothman commented:
The wishlist reference was about using probabilities in chaining. This needs to be done irrespective of which method is called on the Chain

Commit:
doctoring fix

jnothman commented:
As well as the above comments, there are trivial merge conflicts to resolve here. I think we aren't aiming to get this in for the upcoming release in any case.

Commit:
requested changes

Commit:
Merge branch 'master2' into classifier_chain

# Conflicts:
#	sklearn/multioutput.py
#	sklearn/tests/test_multioutput.py

jnothman commented:
flake8 failures

jnothman commented on sklearn/tests/test_multioutput.py:
```diff
@@ -339,3 +348,142 @@ def test_multi_output_exceptions():
+    assert_equal(len(classifier_chain_random.order_), 4)
+    assert_equal(len(set(classifier_chain_random.order_)), 4)
+
+    classifier_chain_fixed = ClassifierChain(LogisticRegression(),
```
a comment like "Random order should behave identically to the same order being specified" before this might be helpful to the reader

jnothman reviewed (approved):
Otherwise, LGTM. Good work!

jnothman commented on sklearn/tests/test_multioutput.py:
```diff
@@ -339,3 +348,142 @@ def test_multi_output_exceptions():
+    Y_decision = classifier_chain.decision_function(X)
+
+    Y_binary = (Y_decision >= 0)
+    assert_array_equal(Y_binary, Y_pred)
```
To be sure can we add `assert not hasattr(classifier_chain, 'predict_proba')`?

adamklec replied:
done

Commit:
requested changes to tests

Commit:
Merge branch 'master2' into classifier_chain

Commit:
removed unused import

Commit:
formatting

Commit:
updated whats_new.rst

adamklec commented on sklearn/utils/testing.py:
```diff
@@ -58,7 +58,6 @@
 
 from sklearn.base import (ClassifierMixin, RegressorMixin, TransformerMixin,
                           ClusterMixin)
-from sklearn.cluster import DBSCAN
```
For mysterious reasons this import was causing trouble on my local version. Since it appears to no longer be used I removed it.

jnothman replied:
Grr... Yes, this is a problem of our CI running flake8 only on diffs. I've removed this import in master.

adamklec commented:
So it seems like we are nearing the end of this journey. :) What are the final steps to get this merged?

jnothman commented:
You need to wait for another reviewer, perhaps @jmschrei, to give it a full review!

jmschrei commented:
I was at  conference for the last week-- I will review this tonight!

jmschrei commented on sklearn/multioutput.py:
```diff
@@ -365,3 +368,232 @@ def score(self, X, y):
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape (n_samples, n_features)
```
Same nitpick here.

jmschrei commented on sklearn/multioutput.py:
```diff
@@ -365,3 +368,232 @@ def score(self, X, y):
+        return Y_pred
+
+    @if_delegate_has_method('base_estimator')
+    def predict_proba(self, X):
```
I think that this could use a little bit more documentation. Specifically, you may want to clarify that `predict` and `predict_proba` don't change the input to the later models in the chain, it always uses `predict.`

jmschrei commented on sklearn/multioutput.py:
```diff
@@ -365,3 +368,232 @@ def score(self, X, y):
+
+    Parameters
+    ----------
+    base_estimator : estimator
```
What are your thoughts on allowing arbitrary lists of classifiers? For instance, one could use a logistic regression for one task, a Random Forest classifier for another, etc.. This is not something that would need to be implemented for this PR to get merged, I'm just asking out of curiosity.

adamklec replied:
Possibly. We'd have to think about how this would work with random chain orders but I could imagine this being a useful feature.

jmschrei commented on examples/multioutput/plot_classifier_chain_yeast.py:
```diff
@@ -0,0 +1,110 @@
+Classifier Chain
+============================
+Example of using classifier chain on a multilabel dataset.
+
```
You may want to elaborate on the connections between classifier chains and model stacking. Might it be worth it to also mention that regression chains are possible? Maybe that's better saved for when someone implements it.

adamklec replied:
I guess my preference is to keep the text concise and focused on the example. If there is a specific point you think we should make here I'd consider adding a sentence or two. Otherwise I'm inclined to leave it as is.

I think we should stay away from dealing with regression models in this PR. But I agree that is a good direction to go.

jmschrei commented on examples/multioutput/plot_classifier_chain_yeast.py:
guaranteed

jmschrei commented:
Overall, like I've said before, this is strong work and I'm glad that we're close to merging it. At the risk of being reviewer 3, I'm wondering if it might be worthwhile to consider the idea of regression chains here, and abstract out a base chain object. @adamklec what are your thoughts? Are regression chains as popular as classification chains?

jnothman commented:
What do you mean by "at the risk of being reviewer 3"?

jnothman commented:
I'm curious, @jmschrei: Do you expect that there is any benefit in waiting until regression chains are implemented before we merge this? I.e. do you expect it will change the API or help find bugs in this implementation?

Here is what I hope to create issues for once this is merged:
* support for multitarget classification
* support for multitarget regression (with a separate `RegressionChain`)
* support for chained features being outputs of `predict_proba` or `decision_function` not just `predict`

jmschrei commented:
When submitting a paper reviewer 3 tends to always be the annoying one that makes you do a bunch of work. I am OK with not having regression chains supported in this PR, I was just interested in discussing it. Otherwise this LGTM once my minor issues are taken care of.

adamklec commented on examples/multioutput/plot_classifier_chain_yeast.py:
indeed

Commit:
fixed docstrings

adamklec commented:
@jmschrei I made the changes you requested. I did not do anything related to regression chains though. That seems like it could be a good idea but I think it should be addressed in another PR.

@jnothman to your list of issues to be addressed I would add
- support for multilabel models in the ensemble module

In `plot_classifier_chain_yeast.py` I implement a very simple voting ensemble of 10 classifier chains. The main problem though is that the 10 chains are trained in series. If `VotingClassifier` supported multilabel targets the chains could be trained in parallel.

jmschrei commented:
Okay. I am in agreement with most of your comments, but will do a full review tonight (and hopefully merge!) I do think it might be valuable somewhere to mention the connection between stacking and classifier chains, since they are similar but not identical. Is there a place you think this might fit?

jnothman commented:
how does support for multilabel ensembles depend on this issue?

On 29 Jun 2017 5:25 am, "Jacob Schreiber" <notifications@github.com> wrote:

> Okay. I am in agreement with most of your comments, but will do a full
> review tonight (and hopefully merge!) I do think it might be valuable
> somewhere to mention the connection between stacking and classifier chains,
> since they are similar but not identical. Is there a place you think this
> might fit?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7602#issuecomment-311762084>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6-joTEln3T7CxWbsYJENZlfZ2m2Zks5sIqijgaJpZM4KRXOp>
> .
>

jmschrei commented:
It doesn't. Like I said, I will review the changes when I get home. In the
meantime I was asking a question.

On Wed, Jun 28, 2017 at 3:51 PM Joel Nothman <notifications@github.com>
wrote:

> how does support for multilabel ensembles depend on this issue?
>
> On 29 Jun 2017 5:25 am, "Jacob Schreiber" <notifications@github.com>
> wrote:
>
> > Okay. I am in agreement with most of your comments, but will do a full
> > review tonight (and hopefully merge!) I do think it might be valuable
> > somewhere to mention the connection between stacking and classifier
> chains,
> > since they are similar but not identical. Is there a place you think this
> > might fit?
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/scikit-learn/scikit-learn/pull/7602#issuecomment-311762084
> >,
> > or mute the thread
> > <
> https://github.com/notifications/unsubscribe-auth/AAEz6-joTEln3T7CxWbsYJENZlfZ2m2Zks5sIqijgaJpZM4KRXOp
> >
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7602#issuecomment-311813960>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADvEEPSRGyowsRH7Nv14LAxP67-Dkflhks5sItjSgaJpZM4KRXOp>
> .
>

jmschrei commented:
LGTM!

jnothman commented:
Congratulations, @adamklec!

On 29 Jun 2017 3:34 pm, "Jacob Schreiber" <notifications@github.com> wrote:

> Merged #7602 <https://github.com/scikit-learn/scikit-learn/pull/7602>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7602#event-1143572673>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6yFUVq21C1iXlz6ciFHyvkyQse7cks5sIzddgaJpZM4KRXOp>
> .
>

adamklec commented:
Awesome! Thanks everyone.
